File: 0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        //heap problem
        //knapsack problem
        //greedy problem
        //sorting
        int n = profits.size();  
        vector<pair<int,int>> store;
        for(int i=0;i<n;i++)
        {
            store.push_back({capital[i],profits[i]});
        }
        sort(store.begin(),store.end());
        priority_queue<int> pq;
        int j = 0;
        for(int i=0; i< k;i++)
        {
            while(j<n && store[j].first <= w)
            {
                pq.push(store[j].second);
                j++;
            }
            if(pq.empty())break;
            w+=pq.top();
            pq.pop();

        }
        return w;
    }
};

File: 0502-ipo/0502-ipo.py
class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        project = sorted([(capital[i],profits[i]) for i in range(len(profits))])
        # project = sorted(zip(capitals,profits))

        pq = []
        i = 0
        while k>0:
            while i < len(project) and project[i][0] <= w:
                heapq.heappush(pq,-project[i][1])
                i+=1
            
            if not pq:
                break
            w+=-heapq.heappop(pq)
            k-=1
        
        return w
